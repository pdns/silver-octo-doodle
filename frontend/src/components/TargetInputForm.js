import React from 'react';
import PropTypes from 'prop-types';
import TextField from '@material-ui/core/TextField';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import InputLabel from '@material-ui/core/InputLabel';
import FormControl from '@material-ui/core/FormControl';
import Button from '@material-ui/core/Button';
import DialogActions from '@material-ui/core/DialogActions';

class TargetInputForm extends React.Component {
  constructor(props) {
    const { defaultHost } = props;
    super(props);
    this.state = {
      fields: {
        name: { value: '', error: false },
        host: { value: defaultHost ? defaultHost.id : '', error: false },
        from: { value: '', error: false },
        to: { value: '', error: false },
      },
    };
  }

  onChange = (event, field) => {
    const newState = { fields: { ...this.state.fields } };
    newState.fields[field] = { value: event.target.value, error: false };
    this.setState(newState);
  }

  onSubmit = () => {
    const newState = { fields: { ...this.state.fields } };
    let atLeastOneError = false;

    Object.keys(this.state.fields).forEach((f) => {
      let { value } = this.state.fields[f];
      if (typeof value === 'string') value = value.trim();
      const error = value.length < 1;
      atLeastOneError = atLeastOneError || error;
      newState.fields[f] = { value, error };
    });

    if (atLeastOneError) {
      this.setState(newState);
    } else {
      const f = newState.fields;
      this.props.onSubmit({
        name: f.name.value,
        hostId: f.host.value,
        from: f.from.value,
        to: f.to.value,
      });
    }
  }

  render() {
    const { fields } = this.state;

    return (
      <form>
        <TextField
          label="Name"
          value={fields.name.value}
          error={fields.name.error}
          onChange={e => this.onChange(e, 'name')}
          margin="normal"
          fullWidth
          required
          autoFocus
        />
        <FormControl
          error={fields.host.error}
          fullWidth
          required
        >
          <InputLabel shrink>Host</InputLabel>
          <Select
            value={fields.host.value}
            onChange={e => this.onChange(e, 'host')}
            displayEmpty
          >
            <MenuItem value='' disabled>Select a host</MenuItem>
            {
              this.props.hosts &&
              this.props.hosts.map((h, i) => <MenuItem key={i} value={h.id}>{h.name}</MenuItem>)
            }
          </Select>
        </FormControl>
        <TextField
          label="From"
          value={fields.from.value}
          error={fields.from.error}
          onChange={e => this.onChange(e, 'from')}
          margin="normal"
          fullWidth
          required
        />
        <TextField
          label="To"
          value={fields.to.value}
          error={fields.to.error}
          onChange={e => this.onChange(e, 'to')}
          margin="normal"
          fullWidth
          required
        />
        <DialogActions>
          <Button color="primary" onClick={this.props.onCancel}>
            Cancel
          </Button>
          <Button variant="contained" color="primary" onClick={this.onSubmit}>
            Create
          </Button>
        </DialogActions>
      </form>
    );
  }
}

TargetInputForm.propTypes = {
  onCancel: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  hosts: PropTypes.array,
  defaultHost: PropTypes.object,
};

export default TargetInputForm;
