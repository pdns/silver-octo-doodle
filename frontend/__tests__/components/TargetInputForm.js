/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import TargetInputForm from '../../src/components/TargetInputForm';

jest.mock('@material-ui/core/TextField', () => 'TextField');
jest.mock('@material-ui/core/Select', () => 'Select');
jest.mock('@material-ui/core/MenuItem', () => 'MenuItem');
jest.mock('@material-ui/core/InputLabel', () => 'InputLabel');
jest.mock('@material-ui/core/FormControl', () => 'FormControl');
jest.mock('@material-ui/core/Button', () => 'Button');
jest.mock('@material-ui/core/DialogActions', () => 'DialogActions');

describe('TargetInputForm', () => {
  it('should render correctly with minimum props', () => {
    const tree = renderer
      .create(<TargetInputForm onCancel={() => ({})} onSubmit={() => ({})} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with hosts prop with no default selection', () => {
    const hosts = [
      { id: '1', name: 'aaa' },
      { id: '2', name: 'bbb' },
    ];
    const tree = renderer
      .create((
        <TargetInputForm
          onCancel={() => ({})}
          onSubmit={() => ({})}
          hosts={hosts}
        />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with hosts prop and default selection', () => {
    const hosts = [
      { id: '1', name: 'aaa' },
      { id: '2', name: 'bbb' },
    ];
    const tree = renderer
      .create((
        <TargetInputForm
          onCancel={() => ({})}
          onSubmit={() => ({})}
          hosts={hosts}
          defaultHost={hosts[1]}
        />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
