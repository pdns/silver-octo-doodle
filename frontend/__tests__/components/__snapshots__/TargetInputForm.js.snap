// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TargetInputForm should render correctly with hosts prop and default selection 1`] = `
<form>
  <TextField
    autoFocus={true}
    error={false}
    fullWidth={true}
    label="Name"
    margin="normal"
    onChange={[Function]}
    required={true}
    value=""
  />
  <FormControl
    error={false}
    fullWidth={true}
    required={true}
  >
    <InputLabel
      shrink={true}
    >
      Host
    </InputLabel>
    <Select
      displayEmpty={true}
      onChange={[Function]}
      value="2"
    >
      <MenuItem
        disabled={true}
        value=""
      >
        Select a host
      </MenuItem>
      <MenuItem
        value="1"
      >
        aaa
      </MenuItem>
      <MenuItem
        value="2"
      >
        bbb
      </MenuItem>
    </Select>
  </FormControl>
  <TextField
    error={false}
    fullWidth={true}
    label="From"
    margin="normal"
    onChange={[Function]}
    required={true}
    value=""
  />
  <TextField
    error={false}
    fullWidth={true}
    label="To"
    margin="normal"
    onChange={[Function]}
    required={true}
    value=""
  />
  <DialogActions>
    <Button
      color="primary"
      onClick={[Function]}
    >
      Cancel
    </Button>
    <Button
      color="primary"
      onClick={[Function]}
      variant="contained"
    >
      Create
    </Button>
  </DialogActions>
</form>
`;

exports[`TargetInputForm should render correctly with hosts prop with no default selection 1`] = `
<form>
  <TextField
    autoFocus={true}
    error={false}
    fullWidth={true}
    label="Name"
    margin="normal"
    onChange={[Function]}
    required={true}
    value=""
  />
  <FormControl
    error={false}
    fullWidth={true}
    required={true}
  >
    <InputLabel
      shrink={true}
    >
      Host
    </InputLabel>
    <Select
      displayEmpty={true}
      onChange={[Function]}
      value=""
    >
      <MenuItem
        disabled={true}
        value=""
      >
        Select a host
      </MenuItem>
      <MenuItem
        value="1"
      >
        aaa
      </MenuItem>
      <MenuItem
        value="2"
      >
        bbb
      </MenuItem>
    </Select>
  </FormControl>
  <TextField
    error={false}
    fullWidth={true}
    label="From"
    margin="normal"
    onChange={[Function]}
    required={true}
    value=""
  />
  <TextField
    error={false}
    fullWidth={true}
    label="To"
    margin="normal"
    onChange={[Function]}
    required={true}
    value=""
  />
  <DialogActions>
    <Button
      color="primary"
      onClick={[Function]}
    >
      Cancel
    </Button>
    <Button
      color="primary"
      onClick={[Function]}
      variant="contained"
    >
      Create
    </Button>
  </DialogActions>
</form>
`;

exports[`TargetInputForm should render correctly with minimum props 1`] = `
<form>
  <TextField
    autoFocus={true}
    error={false}
    fullWidth={true}
    label="Name"
    margin="normal"
    onChange={[Function]}
    required={true}
    value=""
  />
  <FormControl
    error={false}
    fullWidth={true}
    required={true}
  >
    <InputLabel
      shrink={true}
    >
      Host
    </InputLabel>
    <Select
      displayEmpty={true}
      onChange={[Function]}
      value=""
    >
      <MenuItem
        disabled={true}
        value=""
      >
        Select a host
      </MenuItem>
    </Select>
  </FormControl>
  <TextField
    error={false}
    fullWidth={true}
    label="From"
    margin="normal"
    onChange={[Function]}
    required={true}
    value=""
  />
  <TextField
    error={false}
    fullWidth={true}
    label="To"
    margin="normal"
    onChange={[Function]}
    required={true}
    value=""
  />
  <DialogActions>
    <Button
      color="primary"
      onClick={[Function]}
    >
      Cancel
    </Button>
    <Button
      color="primary"
      onClick={[Function]}
      variant="contained"
    >
      Create
    </Button>
  </DialogActions>
</form>
`;
